arithmetization/src/main/java/net/consensys/linea/zktracer/module/shakiradata/ShakiraDataOperation.java

```diff
@@ -0,0 +1,284 @@
+    return (rowIndex, phaseIndex) -> {
+      if (phaseIndex == dataPhase) {
+        final Bytes limbRowData = extractLimbRowData(rowIndex);
+        trace.limb(limbRowData);
+
+        final int nBytesAcc = Math.min((rowIndex + 1) * Trace.LLARGE, hashInput.size());
+        final int nBytes = nBytesAcc - (rowIndex * Trace.LLARGE);
+        trace.nBytes(UnsignedByte.of(nBytes)).nBytesAcc(nBytesAcc);
+
+      } else if (phaseIndex == resultPhase) {
+        trace.limb(computedResult.slice(Trace.LLARGE * rowIndex, Trace.LLARGE));
+        trace.nBytes(UnsignedByte.ZERO).nBytesAcc(0L);
+      }
+    };
+  }
+
+  private static Bytes computeSha256Result(Bytes input) {
+    return Hash.sha256(input);
+  }
+
+  private static Bytes computeKeccakResult(Bytes input) {
+    return Hash.keccak256(input);
+  }
+
+  private static Bytes computeRipeMdResult(Bytes input) {
+    return leftPadTo(Hash.ripemd160(input), 32);
+  }
+
+  private int totalNumberOfDataRows() {
+    return (hashInput.size() + Trace.LLARGEMO) / Trace.LLARGE;
+  }
+
+  private int totalNumberOfRows() {
+    return totalNumberOfDataRows() + RESULT_ROWS_COUNT + EXTRA_ROWS_COUNT;
+  }
+
+  private Map<Integer, PhaseInfo> getPhaseInfoMap() {
+    final int indexMaxDataPhases = totalNumberOfDataRows() - 1;
+    final int totalSizeDataPhases = hashInput.size();
+    final UnsignedByte[] deltaBytesDataPhases = getDataPhaseDeltaBytes();
+    final UnsignedByte[] deltaBytesResultPhases = getResultPhaseDeltaBytes();
+    final UnsignedByte[] deltaBytesExtraPhases = getExtraPhaseDeltaBytes();
+
+    return Map.of(
+        Trace.PHASE_SHA2_DATA,
+        new PhaseInfo(
+            Trace.PHASE_SHA2_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_SHA2_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_SHA2_RESULT,
+            Trace.INDEX_MAX_SHA2_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        Trace.PHASE_RIPEMD_DATA,
+        new PhaseInfo(
+            Trace.PHASE_RIPEMD_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_RIPEMD_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_RIPEMD_RESULT,
+            Trace.INDEX_MAX_RIPEMD_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        Trace.PHASE_KECCAK_DATA,
+        new PhaseInfo(
+            Trace.PHASE_KECCAK_DATA, indexMaxDataPhases, totalSizeDataPhases, deltaBytesDataPhases),
+        Trace.PHASE_KECCAK_RESULT,
+        new PhaseInfo(
+            Trace.PHASE_KECCAK_RESULT,
+            Trace.INDEX_MAX_KECCAK_RESULT,
+            TOTAL_SIZE_RESULT_PHASES,
+            deltaBytesResultPhases),
+        PHASE_EXTRA,
+        new PhaseInfo(
+            PHASE_EXTRA, INDEX_MAX_EXTRA, TOTAL_SIZE_EXTRA_PHASES, deltaBytesExtraPhases));
+  }
+
+  private Bytes extractLimbRowData(Integer rowIndex) {
```

Comment:
An easier way could be to just store the rightPadded (to LLARGE * (input.size +LLARGEMO) / LLARGE) input so after you can do the slicing by LLARGE bytes without checking it at each row (or if you prefer, just check it for the last row, it's useless before)